daemon/device-plugin/deviceplugin.go

```diff
@@ -0,0 +1,243 @@
+	resp := new(pluginapi.AllocateResponse)
+	devName := ""
+	for _, container := range rqt.ContainerRequests {
+		containerResp := new(pluginapi.ContainerAllocateResponse)
+		for _, id := range container.DevicesIDs {
+			fmt.Printf("DeviceID in Allocate: %v \n", id)
+			dev, ok := sm.devices[id]
+			if !ok {
+				fmt.Printf("Error. Invalid allocation request with non-existing device %s", id)
+			}
+			if dev.Health != pluginapi.Healthy {
+				fmt.Printf("Error. Invalid allocation request with unhealthy device %s", id)
+			}
+
+			devName = devName + id + ","
+		}
+
+		fmt.Printf("device(s) allocated: %s\n", devName)
+		envmap := make(map[string]string)
+		envmap["APF-DEV"] = devName
+
+		containerResp.Envs = envmap
+		resp.ContainerResponses = append(resp.ContainerResponses, containerResp)
+	}
+	return resp, nil
+}
+
+func (sm *apfManager) GetDeviceState(DeviceName string) string {
+	// TODO: Discover device health
+	return pluginapi.Healthy
+}
+
+func (sm *apfManager) Start() error {
+
+	sm.ensureConnected()
+
+	ctx := context.Background()
+
+	Devices, err := sm.client.GetDevices(ctx, &emptypb.Empty{})
+
+	for _, device := range Devices.Devices {
+		sm.devices[device.ID] = pluginapi.Device{ID: device.ID, Health: pluginapi.Healthy}
+	}
+
+	for dev := range sm.devices {
+		sm.log.Info(dev)
+	}
+
+	if err != nil {
+		sm.log.Error(err, "Failed to handle GetDevices Request")
+		return err
+	}
+
+	pluginEndpoint := filepath.Join(pluginapi.DevicePluginPath, sm.socketFile)
+	fmt.Printf("Starting APF Device Plugin server at: %s\n", pluginEndpoint)
+	lis, err := net.Listen("unix", pluginEndpoint)
+	if err != nil {
+		fmt.Printf("Error: Starting APF Device Plugin server failed: %v", err)
+	}
+	sm.grpcServer = grpc.NewServer()
+
+	kubeletEndpoint := filepath.Join("unix:", DeprecatedSockDir, KubeEndPoint)
+
+	conn, err := grpc.Dial(kubeletEndpoint, grpc.WithTransportCredentials(insecure.NewCredentials()))
+
+	if err != nil {
+		fmt.Printf("%s device plugin unable connect to Kubelet : %v", resourceName, err)
+		return err
+	}
+	defer conn.Close()
+
+	pluginapi.RegisterDevicePluginServer(sm.grpcServer, sm)
+
+	client := pluginapi.NewRegistrationClient(conn)
+
+	go sm.grpcServer.Serve(lis)
+
+	// Wait for server to start by launching a blocking connection
+	ctx, _ = context.WithTimeout(context.TODO(), 5*time.Second)
+	conn, err = grpc.DialContext(
+		ctx, "unix:"+pluginEndpoint, grpc.WithTransportCredentials(insecure.NewCredentials()), grpc.WithBlock())
+
+	if err != nil {
+		fmt.Printf("error. unable to establish test connection with %s gRPC server: %v", resourceName, err)
+		return err
+	}
+	fmt.Printf("%s device plugin endpoint started serving \n", resourceName)
+	conn.Close()
+
+	ctx = context.Background()
+
+	request := &pluginapi.RegisterRequest{
+		Version:      pluginapi.Version,
+		Endpoint:     sm.socketFile,
+		ResourceName: resourceName,
+	}
+
+	if _, err = client.Register(ctx, request); err != nil {
```

Comment:
what if we call `Start` more than once. Is that an issue?

Reply:
I think it would fail to register, not sure. 
could be fine...

Reply:
if unix socket is not cleaned up - then it will fail, as address will be in use. 
I tested this. 
I have a ``cleanup`` function there, just it needs to be called.

Reply:
currently don't have any call to cleanup - need to check best time to do it

Reply:
I have added this in, currently testing - will commit in a moment.

Reply:
fixed - ensure clean state at start

Reply:
ty. this part lgtm.